### PR TITLE
🩹 fix: Drop Foreign Reasoning Blocks in Anthropic Message Converter

### PR DESCRIPTION
**Pull request by danny-avila**
## Problem

Cross-provider agent handoffs crash with **`Unsupported message content format`** when an agent on **Bedrock Anthropic** (extended thinking enabled) hands off to an agent on the **official Anthropic provider** (e.g. via `lc_transfer_to_agent_*`).

Bedrock represents reasoning as a `reasoning_content` content block:

```jsonc
{ "type": "reasoning_content", "reasoningText": { "text": "...", "signature": "..." } }
```

On handoff, the accumulated history (containing that block) is re-sent through the official-Anthropic converter `_formatContent` (`src/llm/anthropic/utils/message_inputs.ts`), which had no branch for `reasoning_content` and hit the final `else` → `throw new Error('Unsupported message content format')`.

It only manifests on the **streaming** path (`_streamResponseChunks` → `_convertMessagesToAnthropicPayload`), which is the path LibreChat agents always take. The non-streaming `_generate` path uses the base `@langchain/anthropic` formatter, which tolerates the block — which is why this hid in plain sight.

## Fix

- **Drop foreign reasoning blocks** — Bedrock `reasoning_content`, Google `reasoning`, and LibreChat `think`. Their signatures are provider-specific and cannot be validated by Anthropic, so forwarding them is never valid; the receiving model produces its own thinking. (Converting to a `thinking` block would fail Anthropic's signature validation; the existing empty-content placeholder already covers a turn that ends up with no blocks.)
- **Degrade gracefully on any unknown block** — the final `else` now logs and drops instead of throwing, so a single unrecognized block can never crash an entire run again.

## Verification

**Live (real Bedrock → official Anthropic, streaming):** captured a real `reasoning_content` block from `claude-sonnet-4-5` on Bedrock Converse with extended thinking, then sent it through the official Anthropic provider.

| | before | after |
|---|---|---|
| `.stream()` (the agent path) | ❌ `Unsupported message content format` | ✅ normal response |
| `.invoke()` streaming | ❌ same throw | ✅ normal response |

**Unit:** new `cross-provider-reasoning.test.ts` (6 cases — each reasoning family, the unknown-block fallback, signature-not-forwarded, reasoning-only placeholder) + existing `streaming-tool-input.test.ts` all pass. `tsc` build and lint clean.

## Scope

Single-point fix at the provider boundary. Resolves every Bedrock/Google → Anthropic handoff, not just the originally reported configuration.

### CODE REVIEW

**Comment by danny-avila** [comment]
@codex review

File: src/llm/anthropic/utils/message_inputs.ts [hunk 1]
```diff
@@ -720,12 +733,17 @@ function _formatContent(message: BaseMessage) {
...
+        // Unknown block type: log for diagnostics but degrade gracefully by
+        // dropping it instead of throwing, so a single unrecognized block can
+        // never crash an entire run (e.g. a cross-provider handoff).
         console.error(
-          'Unsupported content part:',
+          'Dropping unsupported content part:',
           JSON.stringify(contentPart, null, 2)
         );
-        throw new Error('Unsupported message content format');
+        return null;
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/llm/anthropic/utils/message_inputs.ts:756]: Fixed in 4b5b903. You're right — the graceful drop applied to every role. The fallback is now role-aware: assistant turns drop an unknown block (so a cross-provider handoff never crashes on a provider artifact), but **user/tool turns still throw** `Unsupported message content format`, so unsupported user media (e.g. a `video_url` block on a prompt) is surfaced rather than silently omitted. Foreign reasoning blocks (`reasoning_content`/`reasoning`/`think`) are still dropped regardless of role since they're assistant-internal. Regression test added: `on a user turn, throws on an unsupported block rather than silently dropping it`.

File: src/llm/anthropic/utils/message_inputs.ts [hunk 2]
```diff
@@ -720,12 +733,17 @@ function _formatContent(message: BaseMessage) {
           name: correspondingToolCall.name,
           input: functionCallPart.functionCall.args,
         };
+      } else if (foreignReasoningTypes.some((t) => t === contentPart.type)) {
+        return null;
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/llm/anthropic/utils/message_inputs.ts:756]: Fixed in 4b5b903. Confirmed exactly as described: `convertConverseMessageToLangChainMessage` records the tool only on `message.tool_calls` (message_outputs.ts:181) with `content` holding just the reasoning block, and the array-content path of `_convertMessagesToAnthropicPayload` only `console.warn`'d instead of materializing `tool_calls` — so dropping the reasoning block left the `_` placeholder and silently lost the transfer tool call. The array-content path now materializes any client `tool_calls` not already represented as a `tool_use`/`server_tool_use`/`input_json_delta` block, and strips the lone `_` placeholder when real tool_use blocks are present. Regression test added: `preserves a tool call carried only on tool_calls when its reasoning sibling is dropped`.

**Comment by danny-avila** [comment]
@codex review

File: src/llm/anthropic/utils/message_inputs.ts [hunk 3]
```diff
@@ -809,24 +835,49 @@ export function _convertMessagesToAnthropicPayload(
...
+            .filter(
               (contentPart) =>
-                (contentPart.type === 'tool_use' ||
-                  contentPart.type === 'input_json_delta' ||
-                  contentPart.type === 'server_tool_use') &&
-                contentPart.id === toolCall.id
+                contentPart.type === 'tool_use' ||
+                contentPart.type === 'input_json_delta' ||
+                contentPart.type === 'server_tool_use'
             )
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/llm/anthropic/utils/message_inputs.ts]: Fixed in 2cc7761. You're right — `representedToolIds` was built from the raw content by type, so a `functionCall` part (which `_formatContent` converts into a `tool_use`) wasn't counted as represented, and the materialization then appended a duplicate `tool_use` with the same id. It's now derived from the **formatted output**, so any tool_use `_formatContent` already produced — including from `functionCall` — is recognized as represented and not re-appended. Regression test added: `does not duplicate a Google functionCall tool call already materialized by _formatContent`.

File: src/llm/anthropic/utils/message_inputs.ts [hunk 4]
```diff
@@ -720,7 +733,20 @@ function _formatContent(message: BaseMessage) {
...
+        return null;
+      } else if (isAIMessage(message)) {
+        // Assistant turns can carry provider-specific artifacts (e.g. a new
+        // reasoning variant) that Anthropic cannot consume. Drop them rather
+        // than crash a cross-provider handoff mid-run.
+        console.error(
+          'Dropping unsupported assistant content part:',
+          JSON.stringify(contentPart, null, 2)
+        );
+        return null;
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/llm/anthropic/utils/message_inputs.ts:756]: Fixed in 2cc7761. Agreed — silently dropping `executableCode`/`codeExecutionResult` loses real content the next model needs. I removed the blanket assistant-side drop entirely: only the known foreign reasoning types (`reasoning_content`/`reasoning`/`think`) are dropped, and every other unknown block throws `Unsupported message content format` again (the pre-existing behavior) rather than being silently omitted — for all roles. This also subsumes the round-1 user-media concern uniformly. Regression test added: `throws (not silently drops) on an unknown assistant block such as Google code execution`.

**Comment by danny-avila** [comment]
@codex review

File: src/llm/anthropic/utils/message_inputs.ts [hunk 5]
```diff
@@ -720,6 +733,14 @@ function _formatContent(message: BaseMessage) {
           name: correspondingToolCall.name,
           input: functionCallPart.functionCall.args,
         };
+      } else if (foreignReasoningTypes.some((t) => t === contentPart.type)) {
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/llm/anthropic/utils/message_inputs.ts]: Fixed in 90e5bf1. Agreed — the drop rationale (provider-specific signatures Anthropic can't validate) only applies to assistant artifacts. The foreign-reasoning drop is now gated on `isAIMessage(message)`; a `reasoning_content`/`reasoning`/`think` block on a user or tool turn falls through to the throw and is surfaced rather than silently dropped. Regression test added: `does not drop a reasoning-typed block on a user turn (only assistant reasoning is dropped)`.

**Comment by danny-avila** [comment]
@codex review

**Comment by danny-avila** [comment]
@codex review

File: src/llm/bedrock/utils/message_inputs.ts [hunk 6]
```diff
@@ -644,6 +657,15 @@ function convertAIMessageToConverseMessage(msg: BaseMessage): BedrockMessage {
             type: 'default',
           },
         } as BedrockContentBlock);
+      } else if (FOREIGN_REASONING_TYPES.some((t) => t === block.type)) {
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/llm/bedrock/utils/message_inputs.ts:667]: Fixed in 90bba28. Right — a reasoning-only assistant turn with no tool_calls dropped to empty content, which Bedrock Converse rejects. `convertAIMessageToConverseMessage` now falls back to a placeholder text block when filtering leaves the assistant content empty (mirrors the Anthropic-side `_` placeholder). Regression test added: `emits a placeholder (not empty content) when a reasoning-only turn is fully dropped`.

File: src/llm/anthropic/utils/message_inputs.ts [hunk 7]
```diff
@@ -429,6 +429,14 @@ function _formatContent(message: BaseMessage) {
...
   ];
   const textTypes = ['text', 'text_delta'];
+  /**
+   * Reasoning blocks emitted by other providers — Bedrock's `reasoning_content`,
+   * Google's `reasoning`, and LibreChat's `think`. Their signatures are
+   * provider-specific and cannot be validated by Anthropic, so on a
+   * cross-provider handoff (e.g. Bedrock → Anthropic) we drop them rather than
+   * forwarding an unusable block. The receiving model produces its own thinking.
+   */
+  const foreignReasoningTypes = ['reasoning_content', 'reasoning', 'think'];
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/llm/anthropic/utils/message_inputs.ts:439]: Fixed in 90bba28. Good catch — Google thinking-enabled output uses `type: 'thinking'` with no Anthropic signature (Google's thought-signatures live on functionCall parts in `additional_kwargs`, not on thinking content blocks — confirmed at src/llm/google/llm.spec.ts:1314). The native `thinking` branch now drops unsigned thinking on assistant turns (foreign/Google, which Anthropic rejects) and forwards only signed Anthropic-native thinking. `redacted_thinking` (carries `data`, not a signature) is unaffected. Regression tests added: `drops an unsigned thinking block (Google thinking-enabled output)` and `forwards a signed thinking block (Anthropic-native)`.

**Comment by danny-avila** [comment]
@codex review

File: src/llm/anthropic/utils/message_inputs.ts [hunk 8]
```diff
@@ -808,25 +842,53 @@ export function _convertMessagesToAnthropicPayload(
...
+              'text' in block &&
+              block.text === ANTHROPIC_EMPTY_TEXT_PLACEHOLDER
+            )
+        );
         return {
           role,
-          content: _formatContent(message),
+          content: [
+            ...existingBlocks,
+            ...unrepresentedToolCalls.map(_convertLangChainToolCallToAnthropic),
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/llm/anthropic/utils/message_inputs.ts:890]: Fixed in 725d195. You're right that the pre-pass and the converter disagreed: `ensureThinkingBlockInMessages` counted `reasoning_content` (and unsigned `thinking`/`reasoning`) as a valid thinking block, so it left the cross-provider tool-use turn as an assistant message, and the converter then dropped the reasoning — leaving tool_use with no thinking. The pre-pass is now provider-aware via `isNativeThinkingBlock`: for an Anthropic target only signed `thinking`/`redacted_thinking` count, so a Bedrock `reasoning_content`+tool_use turn is routed to the safe `[Previous agent context]` conversion — the same handling already applied to no-thinking tool-use turns. The 38 existing thinking-handoff tests still pass; 5 cross-provider cases added. Verified end-to-end that the resulting payload is accepted by a thinking-enabled Anthropic agent. (Note: the live API also tolerates a dropped-reasoning tool_use turn with no thinking block, so this is primarily a consistency/defensive alignment with the documented invariant.)

File: src/llm/bedrock/utils/message_inputs.ts [hunk 9]
```diff
@@ -644,6 +664,15 @@ function convertAIMessageToConverseMessage(msg: BaseMessage): BedrockMessage {
...
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/llm/bedrock/utils/message_inputs.ts:667]: Fixed in 725d195 — same provider-aware pre-pass fix, mirrored for Bedrock. `isNativeThinkingBlock` counts only `reasoning_content` (plus `additional_kwargs.reasoning_content`) for a Bedrock target, so an Anthropic `thinking`/`redacted_thinking`+tool_use turn handed to a thinking-enabled Bedrock agent is routed to the safe `[Previous agent context]` conversion instead of being left as a tool-use turn whose reasoning the Bedrock converter then drops. `chainHasThinkingBlock` is provider-aware to match. Covered by the new cross-provider tests + existing Bedrock thinking-chain tests.

**Comment by danny-avila** [comment]
@codex review

File: src/messages/format.ts [hunk 10]
```diff
@@ -1832,12 +1871,7 @@ export function ensureThinkingBlockInMessages(
         }
         if (c.type === 'tool_use') {
           hasToolUse = true;
-        } else if (
-          c.type === ContentTypes.THINKING ||
-          c.type === ContentTypes.REASONING_CONTENT ||
-          c.type === ContentTypes.REASONING ||
-          c.type === 'redacted_thinking'
-        ) {
+        } else if (isNativeThinkingBlock(c, provider)) {
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/messages/format.ts]: Resolved by reverting round 5 in 3307920 — rather than fix the leak inside the conversion path, we removed the conversion for cross-provider foreign-reasoning turns entirely. Live testing confirmed the converter-level drop already yields an Anthropic-accepted payload for a dropped-reasoning tool_use turn (both handoff directions), so these turns now stay as assistant messages (reasoning dropped by the converter), which preserves the tool_use/tool_result structure and avoids serializing reasoning text/signatures into visible context. Thanks — this catch tipped the decision.

File: src/llm/anthropic/utils/message_inputs.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/llm/anthropic/utils/message_inputs.ts:890]: Update: after live-testing, we reverted the round-5 pre-pass change (3307920). The official Anthropic API accepts a dropped-reasoning tool_use turn with no thinking block, so the converter-level drop is sufficient and the pre-pass conversion — which also leaked reasoning into visible context (Finding 10) and discarded tool structure — is unnecessary here. The cross-provider crash this PR targets is fixed by the converter changes; the turn is now left as an assistant message with the foreign reasoning dropped.

File: src/llm/bedrock/utils/message_inputs.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/llm/bedrock/utils/message_inputs.ts:667]: Update: reverted the round-5 pre-pass change for both providers (3307920) — see the Finding 8 and Finding 10 threads for rationale. The Bedrock-side converter drop already handles the Anthropic→Bedrock direction (live-verified); no lossy pre-pass conversion needed.

**Comment by danny-avila** [comment]
@codex review

File: src/llm/anthropic/utils/message_inputs.ts [hunk 11]
```diff
@@ -720,6 +742,18 @@ function _formatContent(message: BaseMessage) {
...
+        foreignReasoningTypes.some((t) => t === contentPart.type)
+      ) {
+        // Foreign reasoning on an ASSISTANT turn (Bedrock `reasoning_content`,
+        // Google `reasoning`, LibreChat `think`) carries provider-specific
+        // signatures Anthropic cannot validate; drop it so a cross-provider
+        // handoff doesn't crash. The same types on a user/tool turn are real
+        // input and fall through to the throw below rather than being silently
+        // dropped — as does any other unknown block (user media, Google
+        // code-execution), which must be surfaced, not discarded.
+        return null;
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/llm/anthropic/utils/message_inputs.ts:756]: This is the same interaction as Findings 8/9, which we resolved by **reverting** the round-5 pre-pass change (3307920) — a deliberate decision after live testing, not an oversight. With the official Anthropic API (claude-sonnet-4-6, thinking enabled), a `[reasoning_content/thinking, tool_use]` history whose foreign reasoning is dropped — i.e. a tool_use turn with no leading thinking block — is **accepted and answered normally**, not rejected. The alternative (routing the turn through `ensureThinkingBlockInMessages`' `[Previous agent context]` conversion) was implemented and reverted because it serialized the foreign reasoning text + signatures into visible context (your Finding 10) and discarded the tool_use/tool_result structure — strictly worse. Distinction that matters here: Anthropic rejects an *invalid/mismatched thinking signature*, not the *absence* of thinking on a historical tool turn. If a concrete repro of rejection surfaces (interleaved-thinking beta, multi-iteration loop), the right fix is a reasoning-**stripped** pre-pass conversion that avoids the Finding-10 leak — but absent that, dropping the foreign block is verified-correct and preserves structure.

File: src/llm/bedrock/utils/message_inputs.ts [hunk 12]
```diff
@@ -644,6 +664,15 @@ function convertAIMessageToConverseMessage(msg: BaseMessage): BedrockMessage {
...
+        // Reasoning from another provider (Anthropic `thinking`/
+        // `redacted_thinking`, Google `reasoning`, LibreChat `think`). Bedrock's
+        // native reasoning is `reasoning_content` (handled above); a foreign
+        // block carries a signature Bedrock cannot validate, so drop it on a
+        // cross-provider handoff (e.g. Anthropic → Bedrock) rather than crash.
+        // The Bedrock model produces its own reasoning. Anything else unknown
+        // still throws below — real content must be surfaced, not dropped.
+        return;
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/llm/bedrock/utils/message_inputs.ts:675]: Same rationale as the Finding 11 thread (mirror direction). We reverted the round-5 provider-aware pre-pass (3307920) deliberately: the converter-level drop produces a payload the live Bedrock/Anthropic thinking APIs accept, while the `[Previous agent context]` conversion leaked foreign reasoning text/signatures into visible context (Finding 10) and lost tool structure. Bedrock's tool calls are materialized from `msg.tool_calls` regardless, so dropping the foreign reasoning block leaves a valid tool turn. If a concrete rejection repro surfaces, the fix would be a reasoning-stripped pre-pass conversion rather than re-introducing the leak.

File: src/llm/anthropic/utils/cross-provider-reasoning.test.ts [hunk 13]
```diff
@@ -0,0 +1,329 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/llm/anthropic/utils/cross-provider-reasoning.test.ts]: The adjacent existing test files in this directory — `streaming-tool-input.test.ts` and `server-tool-inputs.test.ts` — already use this same file-level `/* eslint-disable @typescript-eslint/no-explicit-any */` for the same reason: constructing deliberately cross-provider / edge-case message shapes that don't fit LangChain v1's strict content union, and asserting on loosely-typed provider payloads. AGENTS.md also discourages `unknown` and `Record<string, unknown>`, which leaves no clean typed alternative for these fixtures, so I followed the established sibling-test convention. Happy to convert these two files to narrow local types if you'd prefer the new tests hold a higher bar than the existing ones.

File: src/llm/anthropic/utils/message_inputs.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/llm/anthropic/utils/message_inputs.ts:756]: Settled with live tests (real creds, round-5 reverted, full pipeline ensureThinkingBlockInMessages → converter → provider API):

| scenario | receiver | result |
|---|---|---|
| single tool round, default betas | Anthropic-thinking (claude-sonnet-4-6) | ✅ accepted |
| single round + `interleaved-thinking-2025-05-14` beta | Anthropic-thinking | ✅ accepted |
| multi-round (reasoning on 1st turn only, follow-up tool turn has none) | Anthropic-thinking | ✅ accepted |

In every case the converter drops the foreign `reasoning_content`, leaving an assistant `tool_use` turn with no leading thinking block, and Anthropic accepts and answers normally — it does not reject or degrade. So the converter-level drop is sufficient and the `[Previous agent context]` conversion (which leaked reasoning per Finding 10 and lost tool structure) is not needed here. Holding with the drop.

File: src/llm/bedrock/utils/message_inputs.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/llm/bedrock/utils/message_inputs.ts:675]: Settled with a live test (real creds): Anthropic `thinking`+tool turn → thinking-enabled Bedrock (claude-sonnet-4-5, calculator tool bound). The converter emits a single `toolUse` block (the Anthropic `thinking` is dropped, tool call preserved) and **Bedrock accepts it and answers correctly** ("17 × 23 is 391"). No leading `reasoningContent` is required for a historical tool turn after the foreign block is dropped. Same conclusion as the Finding 11 thread — converter-level drop is sufficient.

File: src/llm/anthropic/utils/cross-provider-reasoning.test.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (danny-avila) [reply src/llm/anthropic/utils/cross-provider-reasoning.test.ts]: Done in d534f64 — converted both files. Removed the file-level `eslint-disable` and every `as any`: content blocks are now plain inline literals (LangChain accepts `{ type, ... }` shapes, so no cast is needed), and assertions read through small explicit view types (`TestBlock` / `ConverseBlock`) plus typed helpers (`assistantBlocks` / `assistantContent`). No `any` remains, no behavior/coverage change — 16 tests pass, build + lint clean.

**Comment by danny-avila** [comment]
@codex review
